internal/db/fetcher/indexer_iterators.go

```diff
@@ -64,13 +64,41 @@ type indexIterResult struct {
 	value    []byte
 }
 
-type queryResultIterator struct {
-	resultIter    query.Results
+// indexPrefixIterator is an iterator over index keys with a specific prefix.
+type indexPrefixIterator struct {
 	indexDesc     client.IndexDescription
 	indexedFields []client.FieldDefinition
+	indexKey      core.IndexDataStoreKey
+	matchers      []valueMatcher
+	execInfo      *ExecInfo
+	resultIter    query.Results
+	ctx           context.Context
+	store         datastore.DSReaderWriter
+}
+
+var _ indexIterator = (*indexPrefixIterator)(nil)
+
+func (iter *indexPrefixIterator) Init(ctx context.Context, store datastore.DSReaderWriter) error {
+	iter.ctx = ctx
+	iter.store = store
+	iter.resultIter = nil
```

Comment:
suggestion: If `iter.resultIter` is not nil here, it looks like it should probably be closed before clearing the ref?

I have a vague memory of the normal fetcher failing because of this some time ago, but I might be wrong.